feat: add configurable typo detection for package dependency constraints

kemsguy7 opened the pull request.

Leonidas-from-XIV commented on src/dune_config/config.ml:
```diff
@@ -86,6 +86,8 @@ let make ~name ~of_string ~default =
 
 let make_toggle ~name ~default = make ~name ~default ~of_string:Toggle.of_string
+
+(* Existing config options *)
```
This comment makes no sense, when reading the code all of the options are existing.

kemsguy7 replied:
Thanks @Leonidas-from-XIV for the feedback, i'm implementing them.

Leonidas-from-XIV commented on src/dune_config/config.ml:
```diff
@@ -147,3 +149,8 @@ let threaded_console_frames_per_second =
     ~default:`Default
 ;;
+
+(* New configuration option for typo warnings *)
```
No need for the comment, this can be determined by `git blame`.

Leonidas-from-XIV commented on src/dune_lang/package_dependency.ml:
```diff
@@ -40,12 +40,64 @@ let encode { name; constraint_ } =
+(* Check for common typos in package dependency constraints *)
+let check_for_typo ~loc { name; constraint_ } =
+  (* Only check for typos if warnings are enabled *)
+  if Dune_config.Config.(get typo_warnings) = `Enabled
```
I would probably flip the condition here so the `None` is the early out option as then you can then mentally concentrate on the other option, without the `None` option hanging out at the very end of the code.

Leonidas-from-XIV commented on src/dune_lang/package_dependency.ml:
```diff
@@ -40,12 +40,64 @@ let encode { name; constraint_ } =
+  (* Only check for typos if warnings are enabled *)
+  if Dune_config.Config.(get typo_warnings) = `Enabled
+  then
+    let open Package_constraint in
```
I'd probably avoid the local open here or if it is this verbose, then use `let module C = Package_constraint in` to avoid potentially pulling in unexpected values.

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/typo-detection.t:
```diff
@@ -0,0 +1,29 @@
+  Warning: Possible typo in constraint for dependency "baz": ':with_test' might
+  be a mistake.
+  Hint: Did you mean to use ':with-test' instead? Example: (depends (baz
+  :with-test))
```
I think it would be nice to also demonstrate that fixing these issues makes the messages go away.

kemsguy7 replied:
@Leonidas-from-XIV , so this particular test was for the first part of this pr : https://github.com/ocaml/dune/pull/11600

The test for this particular PR is in **test/blackbox-tests/test-cases/typo-detection2.t**

gridbugs replied:
You should rebase your branch onto the current tip of origin/main since this file was added in your previous PR which has been merged.

Leonidas-from-XIV reviewed (commented):
Looks overall pretty good, but I have some suggestions that could improve readability of the code.

kemsguy7 commented:
@gridbugs , here's updates I made for issue #11629 .

gridbugs commented on test/blackbox-tests/test-cases/typo-detection2.t:
This should be a cram test like typo-detection.t. Also please choose a more meaningful name for this file.

Leonidas-from-XIV replied:
It probably doesn't get executed because there is no output from `dune pkg lock` as far as I see.

kemsguy7 commented:
@Leonidas-from-XIV @gridbugs , I've created updates for this pr here: https://github.com/ocaml/dune/pull/11766 .  Please check.

Leonidas-from-XIV commented:
@kemsguy7 Why is that a separate PR? It would be better to push your improvements in this PR as then we can refer to the comments that were made in previous reviews.

kemsguy7 commented:
@Leonidas-from-XIV @gridbugs , please kindly review, I've applied updates ass suggested in the comments above

Leonidas-from-XIV commented:
> Sure, please do you mind pointing out the parts that have been unaddressed If you have the time?

Sure. I checked the current state of the PR and these are still unaddressed:

1. [This one](https://github.com/ocaml/dune/pull/11759#discussion_r2073381727)
2. [This one](https://github.com/ocaml/dune/pull/11759#discussion_r2073382279)
3. [This one](https://github.com/ocaml/dune/pull/11759#discussion_r2073384631)
4. [This one](https://github.com/ocaml/dune/pull/11759#discussion_r2073386352)
5. [And this one](https://github.com/ocaml/dune/pull/11759#discussion_r2073390103)

gridbugs commented:
@kemsguy7 the branch this PR is coming from does not include df45715469776a069fda5a4a39daabbb5675a0fe, since it's based on a version of origin/main from before your previous PR was merged. This is why it looks like the changes from your previous PR are duplicated in this PR which is making it hard to review.

Running `git pull origin main --rebase` (and dealing with any conflicts) will update your branch so that all your recent changes are applied to the current tip of origin/main.

kemsguy7 committed:
Implement disabled warnings configuration and fixed merge conflicts with main

Signed-off-by: kemsguy7 <mattidungafa@gmail.com>

kemsguy7 commented:
> @kemsguy7 the branch this PR is coming from does not include [df45715](https://github.com/ocaml/dune/commit/df45715469776a069fda5a4a39daabbb5675a0fe), since it's based on a version of origin/main from before your previous PR was merged. This is why it looks like the changes from your previous PR are duplicated in this PR which is making it hard to review.
> 
> Running `git pull origin main --rebase` (and dealing with any conflicts) will update your branch so that all your recent changes are applied to the current tip of origin/main.

Thank you for this. 
@Leonidas-from-XIV , @gridbugs please kindly review, I've updated with origin main and also implemented changes as requested

Leonidas-from-XIV reviewed (changes requested):
Thanks for the improvements. There's still some things that I think needs adjusting before they can be merged.

In particular you also need to run `dune fmt` to reformat the code to our preferred formatting setting.

kemsguy7 commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,61 +28,62 @@ module Well_formed_name = struct
+  (* Early return if warnings are disabled *)
+  if Dune_config.Config.(get typo_warnings) <> `Enabled then None
+  else
+    let open Package_constraint in
```
Sure

kemsguy7 replied:
@Leonidas-from-XIV please kindly review

kemsguy7 commented on src/dune_config/config.ml:
```diff
@@ -147,3 +147,7 @@ let threaded_console_frames_per_second =
     ~default:`Default
 ;;
+
+let typo_warnings =
```
Do you mean this file config.ml?

gridbugs commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,61 +28,73 @@ module Well_formed_name = struct
   | Some c -> pair Package_name.encode Package_constraint.encode (name, c)
 ;;
 
+(* Check for common typos in package dependency constraints *)
```
This line is duplicated.

kemsguy7 replied:
Fixed

gridbugs commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,61 +28,73 @@ module Well_formed_name = struct
 
   let decode =
-    let open Decoder in
+    let open Dune_sexp.Decoder in
```
Did something change to make this addition necessary now?

Leonidas-from-XIV commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,68 +28,79 @@ module Well_formed_name = struct
 
 let decode =
-  let open Decoder in
+  let open Dune_sexp.Decoder in
```
What change requires this adjustment? I have doubts it is necessary.

kemsguy7 replied:
@Leonidas-from-XIV @gridbugs i've updated this in my latest push

Leonidas-from-XIV replied:
Please check again because the changes are still in place. Check the "files changed" tab in this PR and you'll see that the changes still exist.

Leonidas-from-XIV commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,68 +28,79 @@ module Well_formed_name = struct
+let dependency_constraint_variable_typo_warnings ~loc { name; constraint_ } =
+  (* Handle case where config system isn't initialized yet *)
+  let warnings_enabled =
+    try Dune_config.Config.(get typo_warnings) = `Enabled with
```
Yes, but does `Dune_config.Config.get` throw an exception? If so, which one and why? If so, the `with` case should mention the exception and not just blanket-catch every exception. If not, then this `try` line should be removed.

gridbugs commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,57 +28,68 @@ module Well_formed_name = struct
+    try Dune_config.Config.(get typo_warnings) = `Enabled with
+    | Code_error.E _ -> true (* Only catches Code_error.E exceptions *)
+  in
+  (* Early return if warnings are disabled *)
```
Nitpick about terminology, but this is not an early return. I don't think this comment adds information that isn't obvious reading the code.

gridbugs commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,57 +28,68 @@ module Well_formed_name = struct
+  (* Handle case where config system isn't initialized yet *)
+  let warnings_enabled =
+    try Dune_config.Config.(get typo_warnings) = `Enabled with
+    | Code_error.E _ -> true (* Only catches Code_error.E exceptions *)
```
We definitely shouldn't ever be catching `Code_error` exceptions. The intention of these types of exceptions is to catch programming mistakes inside dune, similar to an assert statement. Did you find that calling `Dune_config.Config.(get typo_warnings)` raised a `Code_error.E` exception?

kemsguy7 replied:
@Leonidas-from-XIV  please can you review the change i did here?  it's making the additional constraint test fail

gridbugs commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,57 +28,68 @@ module Well_formed_name = struct
+    | Code_error.E _ -> true (* Only catches Code_error.E exceptions *)
+  in
+  (* Early return if warnings are disabled *)
+  if not warnings_enabled
```
I recommend avoiding negating the condition of an `if`-statement if it also has an `else` clause (which is almost always the case). Negating the condition is equivalent to swapping the `then` and `else` clauses, and I've found that the latter is easier to understand and less error-prone. If the reason you've structured it this way is because you want the simple case to appear first in the code, consider using a `match` statement instead.

That being said, in this particular case you assign the `warnings_enabled` with the value `Dune_config.Config.(get typo_warnings) = Enabled`. It would be more concise and readable to instead match on `Dune_config.Config.(get typo_warnings)` directly rather than going via an intermediate boolean variable.

kemsguy7 replied:
@gridbugs updated

gridbugs reviewed (changes requested):
Left a few suggestions for improving the readability. Also the current approach to handling exceptions raised by `Dune_config.Config.(get typo_warnings)` doesn't look correct to me.

gridbugs commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,57 +28,68 @@ module Well_formed_name = struct
+  (* Handle case where config system isn't initialized yet *)
+  let warnings_enabled =
+    try Dune_config.Config.(get typo_warnings) = `Enabled with
+    | Code_error.E _ -> true (* Only catches Code_error.E exceptions *)
```
Possibly moot because this line will be removed/changed while applying feedback from this review, but try to avoid comments that don't add any clarity. Assume readers know ocaml syntax, in which case they already understand that `| Code_error.E _ -> ...` in the context of a `try` statement means to catch `Code_error.E` exceptions. A more useful comment would explain why it's necessary to catch such exceptions, or perhaps why other exceptions are not caught. The comment should explain your intent so readers understand it, and maintainers can preserve your intent through changes in the future, or understand the consequence of deleting this code if necessary.

But as I said, it doesn't really matter in this specific instance since I suspect this line will be deleted anyway and as @Leonidas-from-XIV said above, we try to avoid using exceptions as much as possible in dune.

kemsguy7 committed:
fix-update: updated package dependency not to catch exceptions and removed unnecessary comments

Signed-off-by: kemsguy7 <mattidungafa@gmail.com>

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/disabled_warning.t:
```diff
@@ -0,0 +1,60 @@
+  > EOF
+
+  $ DUNE_CONFIG__TYPO_WARNINGS=disabled dune pkg lock
+  $ # No warnings should be produced when the feature is disabled
```
There is no command being run here, so it is missing the test.

Leonidas-from-XIV commented on src/dune_lang/package_dependency.ml:
```diff
@@ -28,57 +28,62 @@ module Well_formed_name = struct
 
 let encode { name; constraint_ } =
-  let open Encoder in
+  let open Dune_sexp.Encoder in
```
Is this change necessary?

kemsguy7 replied:
I'll take it back.

Leonidas-from-XIV commented on src/dune_config/config.mli:
```diff
@@ -60,10 +60,13 @@ val threaded_console : Toggle.t t
     Note that environment variables take precedence over the values passed here
     for easy overriding. *)
-val init : (Loc.t * string) String.Map.t -> unit
+val init : (Loc.t * string) String.Map.t -> unit
```
Please do not remove the trailing newlines. It creates changed lines where there are no changes actually.

Leonidas-from-XIV reviewed (changes requested):
Please make sure to run `dune fmt` to format the source code automatically and take a look at the outstanding review comments.

gridbugs reviewed (changes requested):
Left a few more comments about formatting. Also please make sure to address some of the older comments still outstanding (use the "Files changed" tab to easily see them).
